nest_elephant_tvb/translation/mpi_translator.py

```diff
@@ -3,192 +3,237 @@
     if comm.Get_rank() == 0:
-        bufbytes = buf_pkgs * datasize*2 * pkg_size 
+        bufbytes = datasize * bufsize
     else: 
         bufbytes= 0
     # rank 0: create the shared block
     # rank 1-x: get a handle to it
     win = MPI.Win.Allocate_shared(bufbytes, datasize, comm=comm)
-    buf, datasize = win.Shared_query(0) 
+    buf, datasize = win.Shared_query(0)
     assert datasize == MPI.DOUBLE.Get_size()
     # create a numpy array (buffer) whose data points to the shared mem
-    return np.ndarray(buffer=buf, dtype='d', shape=(buf_pkgs*pkg_size,2))
+    return np.ndarray(buffer=buf, dtype='d', shape=(bufsize,))
 
 
 
-def receive(logger,store,status_data,buffer, comm):
+def _receive(comm_receiver, databuffer, logger):
     '''
-    the receive part of the translator
-    :param logger : logger fro the thread
-    :param store : object for store the data before analysis
-    :param status_data: the status of the buffer (SHARED between thread)
-    :param buffer: the buffer which contains the data (SHARED between thread)
-    :return:
+    Receive data on rank 0. Put it into the shared mem buffer.
+    Replaces the former 'receive' function.
+    NOTE: First refactored version -> not pretty, not final. 
     '''
-    # initialise variables for the loop
-    status_ = MPI.Status() # status of the different message
-    source_sending = np.arange(0,comm.Get_remote_size(),1) # list of all the process for the commmunication
+    status_ = MPI.Status()
+    num_sending = comm_receiver.Get_remote_size() # how many NEST ranks are sending?
+    # TODO: It seems the 'check' variable is used to receive tags from NEST, i.e. ready for send...
+    # change this in the future, also mentioned in the FatEndPoint solution from Wouter.
     check = np.empty(1,dtype='b')
-    count=0
-    while True: # FAT END POINT
-        # send the confirmation of the process can send data
+    shape = np.empty(1, dtype='i')    
+    count = 0
+    # TODO: the last two buffer entries are used for shared information
+    # --> they replace the status_data variable from previous version
+    # --> find more elegant solution?
+    databuffer[-1] = 1 # set buffer to 'ready to receive from nest'
+    databuffer[-2] = 0 # marks the 'head' of the buffer
+    
+    while(True):
         logger.info(" Nest to TVB : wait all")
-        for source in source_sending:
-            comm.Recv([check, 1, MPI.CXX_BOOL], source=source, tag=MPI.ANY_TAG, status=status_)
-
+        head_ = 0 # head of the buffer, reset after each iteration
+         
+        # TODO: This is still not correct. We only check for the Tag of the last rank.
+        # TODO: IF all ranks send always the same tag in one iteration (simulation step)
+        # TODO: then this works. But it should be handled differently!!!!
+        for i in range(num_sending):
+            # new: We do not care which source sends first, give MPI the freedom to send in whichever order.
```

Comment:
I totally agree with this part.
I think that the modification of it will require a redesign of the loop. In my opinion, the best will be the usage of asynchronous call for management of the receiving data. However, I am not sure python can handle correctly this asynchronous call with MPI command.